scripts/migrations/2020-10-archivematica_bags_migration.py

```diff
@@ -0,0 +1,349 @@
+
+        old_bag_info_checksum = existing_checksums.get("bag-info.txt")
+        new_bag_info_checksum = merged_checksums.get("bag-info.txt")
+
+        assert old_bag_info_checksum != new_bag_info_checksum
+
+        with open(f"{working_folder}/{self.tagmanifest_name}", "w") as fp:
+            for checksum, filename in merged_checksums.items():
+                fp.write(f"{filename} {checksum}\n")
+
+    def _upload_bag_to_s3(self, archive_location, working_id, remove_bag=True):
+        s3_upload_key = f"{self.s3_upload_prefix}/{working_id}.tar.gz"
+        workflow_s3_client.upload_file(
+            Filename=archive_location, Bucket=s3_upload_bucket, Key=s3_upload_key
+        )
+        if remove_bag:
+            os.remove(archive_location)
+
+        return s3_upload_key
+
+    def migrate(self, version, space, external_identifier):
+        storage_manifest = storage_client.get_bag(
+            space=space, external_identifier=external_identifier, version=version
+        )
+
+        id = storage_manifest["id"]
+        bucket = storage_manifest["location"]["bucket"]
+        path = storage_manifest["location"]["path"]
+        files = storage_manifest["manifest"]["files"]
+        provider = storage_manifest["location"]["provider"]["id"]
+        tagmanifest_files = storage_manifest["tagManifest"]["files"]
+
+        assert provider == "amazon-s3"
+
+        working_id = id.replace("/", "_")
+        working_folder = f"{self.target_folder}/{working_id}"
+
+        os.makedirs(working_folder, exist_ok=True)
+
+        # Initialise working log
+        with open(f"{self.target_folder}/{working_id}.log", "w") as fp:
+            fp.write(
+                f"{datetime.datetime.now().isoformat()}: Starting migration for {id}\n"
+            )
+
+        def _log(msg):
+            with open(f"{self.target_folder}/{working_id}.log", "a") as fp:
+                fp.write(f"{datetime.datetime.now().isoformat()}: {msg}\n")
+
+        # Write fetch.txt
+        self._write_fetch_file(
+            working_folder=working_folder, bucket=bucket, path=path, files=files
+        )
+        _log(f"Wrote fetch.txt")
+
+        # Get required files from bag
+        self._get_bagit_files_from_s3(
+            working_folder=working_folder,
+            bucket=bucket,
+            path=path,
+            version=version,
+            tagmanifest_files=tagmanifest_files,
+        )
+        _log(f"Got BagIt files from S3")
+
+        # Update bag-info.txt
+        archivematica_uuid = self._get_archivematica_uuid(
+            bucket=bucket, path=path, version=version
+        )
+
+        did_append_uuid = self._append_archivematica_uuid(
+            working_folder, archivematica_uuid
+        )
+        if not did_append_uuid:
+            _log(
+                f"Internal-Sender-Identifier found in bag-info.txt: {archivematica_uuid}"
+            )
+            _log(f"Not migrating {id} (already migrated)")
```

Comment:
Small optimisation: once you've got the bag from the storage-service API, looks for `Internal-Sender-Identifier` in the API response. Saves you doing any downloads from S3 before you skip it here.